doc/src/ds-test-tutorial.md

```diff
@@ -0,0 +1,206 @@
+although this is not explicitly mentioned.
+
+## Test Cases that Must Always Revert
+
+Hevm assumes that a test case should not always revert. If you have such a test
+case, hevm will warn you and return a FAIL. For example this toy contract:
+
+```solidity
+pragma solidity ^0.8.19;
+import {Test} from "forge-std/Test.sol";
+contract MyContract is Test {
+  uint256 cntr;
+  function prove_allrevert(uint256 val) public {
+      if(val < 0) {
+          unchecked { cntr = val; }
+          revert();
+      } else revert();
+  }
+}
+```
+
+When compiled with forge and then ran under hevm with `hevm test`, hevm returns:
+
+```
+Checking 1 function(s) in contract src/contract-allrevert.sol:MyContract
+[RUNNING] prove_allrevert(uint256)
+   [FAIL] prove_allrevert(uint256)
+   Reason:
+     No reachable assertion violations, but all branches reverted
+     Prefix this testname with `proveFail` if this is expected
+```
+
+This is sometimes undesirable. In these cases, prefix your contract with
+`proveFail_` instead of `prove_`:
+
+```solidity
+pragma solidity ^0.8.19;
+import {Test} from "forge-std/Test.sol";
+contract MyContract is Test {
+  uint256 cntr;
+  function proveFail_allrevert_expected(uint256 val) public {
+      if(val < 0) {
+          unchecked {
+            cntr = val;
+            cntr += 1;
+          }
+          revert();
+      }
+      else revert();
+  }
+}
+```
+
+When this is compiled with forge and then checked with hevm, it leads to:
+
+```
+Checking 1 function(s) in contract src/contract-allrevert-expected.sol:MyContract
+[RUNNING] proveFail_allrevert_expected(uint256)
+   [PASS] proveFail_allrevert_expected(uint256)
+```
+
+Which is now the expected outcome.
+
+
+## Supported Cheat Codes
+
+Since hevm is an EVM implementation mainly dedicated to testing and
+exploration, it features a set of `cheat codes` which can manipulate the
+environment in which the execution is run. These can be accessed by calling
+into a contract (typically called `Vm`) at address
+`0x7109709ECfa91a80626fF3989D68f67F5b1DD12D`, which implements the following
+methods
+
+| Function | Description |
+| --- | --- |
+|`function prank(address sender) public`| Sets `msg.sender` to the specified `sender` for the next call.|
+|`function deal(uint usr, uint amt) public`| Sets the eth balance of `usr` to `amt`. Note that if `usr` is a symbolic address, then it must be the address of a contract that has already been deployed. This restriction is in place to ensure soundness of our symbolic address encoding with respect to potential aliasing of symbolic addresses.|
+|`function store(address c, bytes32 loc, bytes32 val) public`| Sets the slot `loc` of contract `c` to `val`.|
+|`function warp(uint x) public`| Sets the block timestamp to `x`.|
+|`function roll(uint x) public`| Sets the block number to `x`.|
+|`function assume(bool b) public`| Add the condition `b` to the assumption base for the current branch. This functions almost identically to `require`.|
```

Comment:
Actually, I also am not 100% sure -- it was worded like that before, so I left it as-is. Looking at the code:

```
  , action "assume(bool)" $
      \sig _ _ input -> case decodeStaticArgs 0 1 input of
        [c] -> modifying #constraints ((:) (PEq c (Lit 1)))
        _ -> vmError (BadCheatCode sig)
```

It looks to me indeed very similar in behaviour. The difference is likely that `require` is built-in and so solidity will be able to understand and maybe optimize it, potentially even optimizing it away. However, `require` is a call out to another contract, which Solidity cannot "see inside", so e.g. it cannot be optimized away.

Reply:
For the average developer, the two are incredibly similar, and I'd recommend using `require`. I'll probably add that to the text. Thanks for poking at this! :)

Reply:
I hope it's good now. I am closing because I think I managed to explain this more thoroughly to the potential users. Let me know if it's not OK, and then I'll improve :) The final text is:


Add the condition `b` to the assumption base for the current branch. This functions almost identically to `require`. For most users, `require` is preferable. However, in case you wish to understand & modify the internal IR of hevm, you may want to use `assume`.